### PR TITLE
Pack QUIC_SENT_PACKET_METADATA manually to remove the need for pragma pack

### PR DESCRIPTION
**Pull request by ThadHouse**
This causes everything to be aligned correctly, removing potential bugs and performance penalties for unaligned reads and writes.

This commit currently loses 1 byte per packet, but gains 4 per frame. However, with a future optimization for StreamOffset, we can actually lose 4 bytes per frame, which will be a net win.

Solves #806 and many more silent packed bugs.

### CODE REVIEW

File: src/core/sent_packet_metadata.h [hunk 1]
```diff
@@ -105,12 +106,21 @@ typedef struct QUIC_SENT_PACKET_METADATA {
     //
     // Frames included in this packet.
     //
-    uint8_t FrameCount;
+    uint8_t FrameCount  : 4;
+    uint8_t PathId      : 4;
```
Comment (nibanks) [inline src/core/sent_packet_metadata.h:110]: This isn't going to work. PathId is a constantly rotating value. We might have a max of X paths at one time, but paths come and go and this value constantly increases.

File: src/core/sent_packet_metadata.h [hunk 2]
```diff
@@ -67,6 +59,16 @@ typedef struct QUIC_SENT_FRAME_METADATA {
...
+    //
+    // The following to fields are for STREAM. However, if they were in stream
+    // they force the union to completely contain them, which doesn't allow the
+    // Type and Flags fields to be packed nicely.
+    //
+    //
+    // TODO- optimization: encode in 32 bits.
+    //
+    uint64_t StreamOffset;
+    uint16_t StreamLength;
```
Comment (nibanks) [inline src/core/sent_packet_metadata.h:71]: Really, we only need ~12 bits here if we assert max packet size is < 4096.

**Comment by nibanks** [comment]
If this is really the direction we want to go to fix these unaligned read/writes we should just try removing the packed as is and see how memory usage changes in tests like RPS. See what peak RPS is for 16K payload before and after this change.

**Comment by ThadHouse** [comment]
> If this is really the direction we want to go to fix these unaligned read/writes we should just try removing the packed as is (no other changes for now, or at least no extra bit squashing) and see how memory usage changes in tests like RPS. See what peak RPS is for 16K payload before and after this change.

Sounds good. I'll leave the StreamOffset change in, since that isn't bit packing but does actually make a large noticeable difference.

**Review by nibanks — APPROVED** [review]

**Comment by ThadHouse** [comment]
This looks to be well within the margin of error for perf. Assuming we can do the 32 bit optimization for offset, we'll even gain a lot back. But for now this looks to not hurt perf. Which makes sense, since the cost is 4 bytes per frame.

**Comment by ThadHouse** [comment]
Just to leave as a note, the only packed struct remaining in production source code is the QUIC packet headers themselves, which need to be packed on purpose. There are a few in tests, and a few in the ETW logging, but otherwise we're back at a safe point.

**Pull request merged**